### ISSUE TITLE
feature: dynamic block pruning for attention (cauchy-schwarz bounds)

### ISSUE THREAD

**Issue by teerthsharma** [geohot/tinygrad#14247]
proposal to implement geometric block pruning for attention.

currently, we compute $N^2$ interactions even for topologically disjoint queries/keys.
this implements a bound check to skip blocks that are mathematically guaranteed to result in near-zero attention scores.

similar to block-sparsity techniques used in TensorRT-LLM for long context inference.

### mechanism
we use the cauchy-schwarz inequality to upper-bound the dot product for a block of keys:
`max(q·k) <= q·μ + norm(q)·r`

if `upper_bound < threshold`, the entire block is masked out before the heavy matmul.
tinygrad's lazy graph handles this naturally (ops don't execute if masked early).

### flow
```mermaid
graph TD
    K[Keys] -->|Blockify| B[Key Blocks]
    B -->|Precompute| G[Centroids & Radii]
    Q[Query] -->|Compare| G
    G -->|Bound Check| D{Bound < Threshold?}
    D -- Yes --> S[Skip Block]
    D -- No --> C[Compute Attention]

**Comment by teerthsharma** [comment]
def geometric_mask(q, k, block_size=64, threshold=0.1):
  # 1. precompute geometry (centroids + radii)
  # we treat a block of keys as a hypersphere.
  # centroid (μ) = mean of block
  # radius (r) = max euclidean distance from centroid to any key in block
  k_blocks = k.reshape(k.shape[0], -1, block_size, k.shape[-1])
  centroids = k_blocks.mean(axis=2)
  radii = ((k_blocks - centroids.unsqueeze(2))**2).sum(axis=3).sqrt().max(axis=2)

  # 2. bound check using cauchy-schwarz inequality
  # we want to skip if max(q·k) < threshold
  # derivation:
  #   q·k = q·(μ + δ), where δ = k - μ (displacement from centroid)
  #   q·k = q·μ + q·δ
  # apply cauchy-schwarz to error term: q·δ <= ||q||·||δ||
  # since ||δ|| <= radius (r), we establish the upper bound:
  #   max(q·k) <= q·μ + ||q||·r
  
  q_norm = (q**2).sum(axis=2).sqrt().unsqueeze(2)
  bound = (q.unsqueeze(2) * centroids.unsqueeze(1)).sum(axis=3) + q_norm * radii.unsqueeze(1)

  # 3. return mask (1=keep, 0=skip)
  # if theoretical upper bound < threshold, actual dot product is guaranteed irrelevant.
  return (bound > threshold).float()

**Comment by teerthsharma** [comment]
### complexity analysis & gradient flow

adding some math context on why this doesn't kill training stability.

**1. forward complexity**
standard attention is $\mathcal{O}(N^2)$.
this implementation is $\mathcal{O}(N \cdot k)$ where $k$ is the number of active blocks.
the overhead for computing centroids/radii is $\mathcal{O}(N)$ (linear scan over keys), which is amortized across all query heads.

**2. backward pass (the important part)**
since `mask` is a hard gate ($1$ or $0$):
$$\frac{\partial \mathcal{L}}{\partial \text{Block}_i} = \frac{\partial \mathcal{L}}{\partial \text{Attn}} \cdot \mathbb{1}_{\text{mask}}$$
if a block is pruned in forward, its gradient is strictly zero. we don't need to compute dense gradients and then mask them. the graph itself is sparse. this is similar to how ReLU kills gradients in the negative regime, but applied to spatial blocks.

**3. relation to tensorRT-llm**
this is functionally equivalent to the "Skip Softmax" optimization in TensorRT-LLM, but implemented as a high-level graph rewrite rather than a custom kernel.
we trade a small amount of "safe" precision (defined by $\tau$) for massive IO reduction.

**4. bounding rigor**
we use the triangle inequality to ensure the radius $r$ is a strict upper bound.
$$\max(q \cdot k) \le q \cdot \mu + ||q|| \cdot r$$
this holds for all $k$ in the block. if the RHS < $\tau$, there is mathematically zero chance of a high attention score. we aren't guessing; we are proving it's safe to skip.

**Comment by teerthsharma** [comment]
(in hopes it doesnt get closed for no reason this time)
